### PR TITLE
Add support for bindPort2 parameter for .NET

### PR DESCRIPTION
**Pull request by qianz2**
- [ ] The purpose of this PR is explained in this message or in an issue. If an issue please include a reference as \#\<issue\_number\>.
- [ ] Tests are included and/or updated for code changes.
- [ ] Proper license headers are included in each file.

Oryx side change:
•        Surface a new optional -bindPort2 parameter.
•        When this parameter is set, pass the following env variables to the .NET 7 process:
o   Kestrel__Endpoints__Http2__Url=http://*:8585
(Port value here comes from the -bindPort2 parameter)
o   Kestrel__Endpoints__Http2__Protocols=Http2
o   Kestrel__Endpoints__Http__Url=http://*:8080
(Port value here comes from the -bindPort parameter)
o   Kestrel__Endpoints__Http__Protocols=Http1
•        We won’t change the existing behavior if -bindPort2 is not provided.

### CODE REVIEW

File: src/startupscriptgenerator/src/dotnetcore/main.go [hunk 1]
```diff
@@ -41,6 +41,7 @@ func main() {
 		"[Optional] Path to the directory where build manifest file can be found. If no value is provided, then "+
 			"it is assumed to be under the directory specified by 'appPath'.")
 	bindPortPtr := scriptCommand.String("bindPort", "", "[Optional] Port where the application will bind to. Default is 8080")
+	bindPort2Ptr := scriptCommand.String("bindPort2", "", "[Optional] Port where the application will bind to. Default is 8080")
```
Comment (daniv-msft) [inline src/startupscriptgenerator/src/dotnetcore/main.go:44]: I know it's only a draft PR, so you would have probably fixed it anyway, but here the description should be updated. I'd recommend something like: "[Optional] .NET only - HTTP/2 port where the application will bind to. By default, no binding will be made if this is not set"

Comment (qianz2) [reply src/startupscriptgenerator/src/dotnetcore/main.go]: Updated

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go [hunk 2]
```diff
@@ -69,6 +69,39 @@ func (gen *DotnetCoreStartupScriptGenerator) shouldApplicationInsightsBeConfigur
...
+	
+	if dotNetRuntimeVersion != "" {
+		dotNetRuntimeVersionConstraint, err := semver.NewConstraint(">= " + versionConstraint)
+		if err != nil {
+    		fmt.Printf("\nError in creating semver constraint %s", err)
+		}
+
+		dotNetCurrentVersion, err := semver.NewVersion(dotNetRuntimeVersion)
+		if err != nil {
+    		fmt.Printf("\nError in parsing current version to semver version %s", err)
```
Comment (cormacpayne) [inline src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:92]: nit: same two comments here as above

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go [hunk 3]
```diff
@@ -69,6 +69,39 @@ func (gen *DotnetCoreStartupScriptGenerator) shouldApplicationInsightsBeConfigur
...
+		}
+		// Check if the version meets the constraints. The a variable will be true.
+		constraintCheckResult := dotNetRuntimeVersionConstraint.Check(dotNetCurrentVersion)
+				
+	    if constraintCheckResult  {
+			   fmt.Printf("\nBefore returning true")
```
Comment (cormacpayne) [inline src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:98]: nit: can we remove this message? In addition, if we don't need to log any additional information, then I'd replace this `if constraintCheckResult` block with the following:

```go
return constraintCheckResult
```

**Review by cormacpayne — DISMISSED** [review]
Minor comments, otherwise LGTM :shipit:

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go (hunk 3 above)
Comment (qianz2) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go]: Updated

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go [hunk 4]
```diff
@@ -69,6 +69,39 @@ func (gen *DotnetCoreStartupScriptGenerator) shouldApplicationInsightsBeConfigur
...
+	} else {
+		dotNetRuntimeVersion = os.Getenv("DOTNET_VERSION")
+	}
+	
+	fmt.Printf("\nDotNet Runtime %s", dotNetRuntimeVersion)
+	
+	if dotNetRuntimeVersion != "" {
+		dotNetRuntimeVersionConstraint, err := semver.NewConstraint(">= " + versionConstraint)
+		if err != nil {
+    		fmt.Printf("\nError in creating semver constraint %s", err)
```
[the remark this answers is not in the document]
Comment (qianz2) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:87]: Added the return false here. For the indentation not sure why github shows differently with what I see from VS code:
Screenshot 2022-10-05 144922

[the remark this answers is not in the document]
Comment (cormacpayne) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:87]: Hmm I've seen GitHub do this before -- probably some weird tab vs. spaces display thing in their UI 🤔

**Review by cormacpayne — DISMISSED** [review]

File: src/startupscriptgenerator/src/dotnetcore/main.go [hunk 5]
```diff
@@ -41,6 +41,7 @@ func main() {
 		"[Optional] Path to the directory where build manifest file can be found. If no value is provided, then "+
 			"it is assumed to be under the directory specified by 'appPath'.")
 	bindPortPtr := scriptCommand.String("bindPort", "", "[Optional] Port where the application will bind to. Default is 8080")
+	bindPort2Ptr := scriptCommand.String("bindPort2", "", "[Optional] .NET only - HTTP/2 port where the application will bind to. By default, no binding will be made if this is not set")
```
Comment (daniv-msft) [inline src/startupscriptgenerator/src/dotnetcore/main.go:44]: Minor comment: I know I'm the one who suggested that particular sentence, but please consider updating to: `By default, no HTTP/2 binding will be made`.
If this is the only comment preventing you from merging, not a big deal, feel free to ignore.

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go [hunk 6]
```diff
@@ -81,6 +113,15 @@ func (gen *DotnetCoreStartupScriptGenerator) GenerateEntrypointScript(scriptBuil
 	common.SetEnvironmentVariableInScript(scriptBuilder, "PORT", gen.BindPort, DefaultBindPort)
 	scriptBuilder.WriteString("export ASPNETCORE_URLS=http://*:$PORT\n\n")
 
+	logger.LogInformation("Setting up Kestrel Endpoints with BindPort and BindPort2 Env variables if .NET runtime version >= 7")
```
Comment (daniv-msft) [inline src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:116]: I actually don't believe we need to restrict it to .NET 7. I believe the info in the email was to limit it to .NET 7 if it was easier, but that fundamentally we wanted it for all .NET versions. Could you please clarify: is there any way to limit it to .NET 7?

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go [hunk 7]
```diff
@@ -81,6 +113,15 @@ func (gen *DotnetCoreStartupScriptGenerator) GenerateEntrypointScript(scriptBuil
...
+	if gen.isDotnetRuntimeVersionMeetConstraint("7.0.0-0") {
+		scriptBuilder.WriteString("if [ -z \"$" + gen.BindPort2 + "\" ]; then" + "\n")
+		scriptBuilder.WriteString("		export Kestrel__Endpoints__Http2__Url=http://*:" + gen.BindPort2 + "\n")
+		scriptBuilder.WriteString("		export Kestrel__Endpoints__Http2__Protocols=Http2\n")
+		scriptBuilder.WriteString("		export Kestrel__Endpoints__Http1__Url=http://*:$PORT\n\n")
```
Comment (daniv-msft) [inline src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:121]: The $PORT value is actually expected to be replaced by the value provided in the BindPort parameter, if I remember correctly.

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go [hunk 8]
```diff
@@ -81,6 +113,15 @@ func (gen *DotnetCoreStartupScriptGenerator) GenerateEntrypointScript(scriptBuil
 	common.SetEnvironmentVariableInScript(scriptBuilder, "PORT", gen.BindPort, DefaultBindPort)
 	scriptBuilder.WriteString("export ASPNETCORE_URLS=http://*:$PORT\n\n")
 
+	logger.LogInformation("Setting up Kestrel Endpoints with BindPort and BindPort2 Env variables if .NET runtime version >= 7")
+	if gen.isDotnetRuntimeVersionMeetConstraint("7.0.0-0") {
+		scriptBuilder.WriteString("if [ -z \"$" + gen.BindPort2 + "\" ]; then" + "\n")
```
Comment (daniv-msft) [inline src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:118]: We should also validate that BindPort is set as well, as we'll need to read its value to set Kestrel__Endpoints__Http1__Url

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go (hunk 7 above)
Comment (daniv-msft) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:121]: Oh my bad, I actually just saw that we were doing this in the file before:
`common.SetEnvironmentVariableInScript(scriptBuilder, "PORT", gen.BindPort, DefaultBindPort)`
So the env variable should already be set properly there.

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go (hunk 8 above)
Comment (daniv-msft) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:118]: Or the $PORT value if it's easier, as it's already defined above?

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go (hunk 6 above)
Comment (qianz2) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:116]: Yes we're actually limited to .NET 7 with the function at line 118. My understanding from the email thread they're asking to scope to .NET 7 first, then .NET 6.

`if gen.isDotnetRuntimeVersionMeetConstraint("7.0.0-0") {`

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go (hunk 8 above)
Comment (qianz2) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:118]: The $PORT value was setup here, and it has default value of 8080:
`	common.SetEnvironmentVariableInScript(scriptBuilder, "PORT", gen.BindPort, DefaultBindPort)
`

File: src/startupscriptgenerator/src/dotnetcore/main.go (hunk 5 above)
Comment (qianz2) [reply src/startupscriptgenerator/src/dotnetcore/main.go]: Updated

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go (hunk 6 above)
Comment (qianz2) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:117]: Or we should remove this line? Please let me know if I understand it wrongly.

Comment (daniv-msft) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:117]: Yes, it's ambiguous. Jenny and I mentioned that we should support all .NET versions if it's possible, but Surender mentioned that having only .NET 7 would be better to reduce the blast radius.

Let's keep what you have, and we can remove the line as needed in the future.

File: src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go (hunk 8 above)
Comment (daniv-msft) [reply src/startupscriptgenerator/src/dotnetcore/scriptgenerator.go:119]: OK fair enough! Thanks for clarifying

**Review by daniv-msft — APPROVED** [review]

File: src/BuildScriptGenerator/DefaultDockerfileGenerator.cs [hunk 9]
```diff
@@ -70,6 +70,11 @@ public string GenerateDockerfile(DockerfileContext ctx)
                     createScriptArguments.Add("bindPort", this.commonOptions.BindPort);
                 }
 
+                if (!string.IsNullOrEmpty(this.commonOptions.BindPort2))
```
Comment (surenderssm) [inline src/BuildScriptGenerator/DefaultDockerfileGenerator.cs:73]: NIT: Checking for IsNullOrWhiteSpace might add more value here.

**Pull request merged**